students/yushus/session07/html_render.py

```diff
@@ -0,0 +1,123 @@
+#!/usr/bin/env/python3
+
+"""
+Yushu Song
+Au2018-Py210B
+HTML Renderer assignment
+"""
+
+import os
+import random
+import re
+import sys
+
+from collections import defaultdict
+
+# This is the framework for the base class
+class Element(object):
+    tag = "html"
+    indent = "    "
+
+    def __init__(self, content=None, **attributes):
+        self.contents = [content] if content else []
+        self.attributes = attributes
+    
+    def _open_tag(self):
+        return f"<{self.tag}{self._get_attributes()}>"
+
+    def _close_tag(self):
+        return f"</{self.tag}>"
+
+    def _get_attributes(self):
+        page_attributes = ""
+        if self.attributes:
+            for attribute in self.attributes:
+                page_attributes = f"{page_attributes} {attribute}=\"{self.attributes[attribute]}\""
```

Comment:
Minor: You are performing string concatenation using something similar to `s = s + append_stuff', which works, but not recommended. The desirable approach is to create a list of component strings and then use string 'join()' method. In fact, it can all be done in one line using list comprehension:
```
return ' '.join(f'{k}={v}' for k, v in self.attributes.items())
```

Reply:
Nice! Thanks Hosung!